diff --git configs/zynq_zybo_defconfig configs/zynq_zybo_defconfig
index 7c23fec..1dead98 100644
--- configs/zynq_zybo_defconfig
+++ configs/zynq_zybo_defconfig
@@ -22,3 +22,8 @@ CONFIG_DEBUG_UART_ZYNQ=y
 CONFIG_DEBUG_UART_BASE=0xe0001000
 CONFIG_DEBUG_UART_CLOCK=50000000
 CONFIG_ZYNQ_QSPI=y
+# CONFIG_OF_SEPARATE is not set
+CONFIG_OF_EMBED=y
+CONFIG_MENU=y
+CONFIG_MENU_SHOW=y
+CONFIG_CMD_BOOTMENU=y
diff --git include/configs/zynq-common.h include/configs/zynq-common.h
index 982905d..dde516a 100644
--- include/configs/zynq-common.h
+++ include/configs/zynq-common.h
@@ -215,7 +215,13 @@
 	"nor_flash_off=0xE2100000\0"	\
 	"fdt_high=0x20000000\0"		\
 	"initrd_high=0x20000000\0"	\
-	"norboot=echo Copying FIT from NOR flash to RAM... && " \
+	"loadbootenv_addr=0x2000000\0" \
+	"fdt_addr_r=0x1f00000\0"        \
+	"pxefile_addr_r=0x2000000\0"    \
+	"kernel_addr_r=0x2000000\0"     \
+	"scriptaddr=0x3000000\0"        \
+	"ramdisk_addr_r=0x3100000\0"    \
+	"norboot=echo Copying FIT from NOR flash to RAM... && "              \
 		"cp.b ${nor_flash_off} ${load_addr} ${fit_size} && " \
 		"bootm ${load_addr}\0" \
 	"sdboot=echo Copying FIT from SD to RAM... && " \
@@ -227,10 +233,50 @@
 	"usbboot=if usb start; then " \
 			"echo Copying FIT from USB to RAM... && " \
 			"load usb 0 ${load_addr} ${fit_image} && " \
-			"bootm ${load_addr}; fi\0" \
-		DFU_ALT_INFO
-
-#define CONFIG_BOOTCOMMAND		"run $modeboot"
+			"bootm ${load_addr}\0" \
+		"fi\0" \
+	"bootenv=uEnv.txt\0" \
+	"bootenv_dev=mmc\0" \
+        "config_name=" CONFIG_SYS_CONFIG_NAME "\0" \
+	"loadbootenv=load ${bootenv_dev} 0 ${loadbootenv_addr} ${bootenv}\0" \
+	"importbootenv=echo Importing environment from ${bootenv_dev} ...; " \
+		"env import -t ${loadbootenv_addr} $filesize\0" \
+	"bootenv_existence_test=test -e ${bootenv_dev} 0 /${bootenv}\0" \
+	"setbootenv=if env run bootenv_existence_test; then " \
+			"if env run loadbootenv; then " \
+				"env run importbootenv; " \
+			"fi; " \
+		"fi; \0" \
+	"sd_loadbootenv=set bootenv_dev mmc && " \
+			"run setbootenv \0" \
+        "loadbootscript=load ${bootenv_dev} 0 ${load_addr} boot.scr\0" \
+        "bootscript=echo Running bootscript from mmc ...;" \
+                "source ${load_addr}\0" \
+	DFU_ALT_INFO
+
+#define CONFIG_PREBOOT \
+	"if test $modeboot = sdboot; then " \
+		"if mmc rescan; then " \
+			"run sd_loadbootenv; " \
+		"fi;" \
+	"fi;"
+
+#define CONFIG_BOOTCOMMAND \
+	"if mmc rescan; then " \
+		"if test -n $uenvcmd; then " \
+			"echo Running uenvcmd ...;" \
+			"run uenvcmd;" \
+		"fi;" \
+		"if run loadbootscript; then " \
+			"run bootscript; " \
+		"fi; " \
+	"fi;" \
+        "run $modeboot"
+
+#define CONFIG_MENU
+#define CONFIG_MENU_SHOW
+#define CONFIG_CMD_BOOTMENU
+#define CONFIG_CMD_BOOTZ
 #define CONFIG_BOOTDELAY		3 /* -1 to Disable autoboot */
 #define CONFIG_SYS_LOAD_ADDR		0 /* default? */
 
